### PR TITLE
add explaination of hairpin problem

### PR DESCRIPTION
**Pull request by freehan**
What do you think about adding more detail?  @bprashanth

### CODE REVIEW

**Comment by bprashanth** [comment]
Great idea!

**Comment by freehan** [comment]
Too much time spent on explaining this to every customer who bump into this.

File: docs/user-guide/debugging-services.md [hunk 1]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
...
+
+* Ensure `--hairpin-mode` is set to `haripin-veth` or `promiscuous-bridge`.
+
+* Check if there is any log lines with key word `hairpin` in kubelet.log. The log is usually
+located at `/var/log/kubelet.log` on node. There should be log lines indicating the effective
+hairpin mode. Please be noted that the effective hairpin mode may not match `--hairpin-mode` flag.
+
+* If the effective hairpin mode is `hairpin-veth`, ensure `Kubelet` has the permission to operate in `/sys` on node.
+
+* If the effective hairpin mode is `promiscuous-bridge`, ensure `Kubelet` has the permission to manipulate linux bridge on node.
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: also tell them to netstat -i on the bridge and give an example please

File: docs/user-guide/debugging-services.md [hunk 2]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
...
+
+The common steps to trouble shoot this are as follows:
+
+* Ensure `--hairpin-mode` is set to `haripin-veth` or `promiscuous-bridge`.
+
+* Check if there is any log lines with key word `hairpin` in kubelet.log. The log is usually
+located at `/var/log/kubelet.log` on node. There should be log lines indicating the effective
+hairpin mode. Please be noted that the effective hairpin mode may not match `--hairpin-mode` flag.
+
+* If the effective hairpin mode is `hairpin-veth`, ensure `Kubelet` has the permission to operate in `/sys` on node.
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: tell: `for intf in /sys/devices/virtual/net/cbr0/brif/*; do cat $intf/hairpin_mode; done`

File: docs/user-guide/debugging-services.md [hunk 3]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
...
+This mostly happens in `iptables` mode. There is `--hairpin-mode` flag in `Kubelet`.
+In order to allow endpoints of a Service to loadbalance back to themselves
+if they should try to access their own Service, `--hairpin-mode` flag has to be set
+to `haripin-veth` or `promiscuous-bridge`.
+
+The common steps to trouble shoot this are as follows:
+
+* Ensure `--hairpin-mode` is set to `haripin-veth` or `promiscuous-bridge`.
+
+* Check if there is any log lines with key word `hairpin` in kubelet.log. The log is usually
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: what about if it's none? (impact, and how to fix, or link to doc, or ask them to contact us or something)

File: docs/user-guide/debugging-services.md [hunk 4]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
...
+### A Pod cannot reach itself via Service IP
+
+This mostly happens in `iptables` mode. There is `--hairpin-mode` flag in `Kubelet`.
+In order to allow endpoints of a Service to loadbalance back to themselves
+if they should try to access their own Service, `--hairpin-mode` flag has to be set
+to `haripin-veth` or `promiscuous-bridge`.
+
+The common steps to trouble shoot this are as follows:
+
+* Ensure `--hairpin-mode` is set to `haripin-veth` or `promiscuous-bridge`.
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: how do they do this?

File: docs/user-guide/debugging-services.md [hunk 5]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
   Engine doesn't do port remapping, so if your application serves on 8080,
   the `containerPort` field needs to be 8080.
 
+### A Pod cannot reach itself via Service IP
+
+This mostly happens in `iptables` mode. There is `--hairpin-mode` flag in `Kubelet`.
+In order to allow endpoints of a Service to loadbalance back to themselves
+if they should try to access their own Service, `--hairpin-mode` flag has to be set
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: does this apply regardless of network plugin ? (not for non-bridge)

File: docs/user-guide/debugging-services.md [hunk 6]
```diff
@@ -555,6 +555,25 @@ There are three things to check:
   Engine doesn't do port remapping, so if your application serves on 8080,
   the `containerPort` field needs to be 8080.
 
+### A Pod cannot reach itself via Service IP
+
+This mostly happens in `iptables` mode. There is `--hairpin-mode` flag in `Kubelet`.
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: what is in iptables-mode?

**Comment by freehan** [comment]
Updated. PTAL

File: docs/user-guide/debugging-services.md [hunk 7]
```diff
@@ -555,6 +555,63 @@ There are three things to check:
...
+to access logs. Please be noted that the effective hairpin mode may not
+match `--hairpin-mode` flag due to compatibility. Check if there is any log
+lines with key word `hairpin` in kubelet.log. There should be log lines
+indicating the effective hairpin mode, like something below.
+
+```shell
+I0629 00:51:43.648698    3252 kubelet.go:380] Hairpin mode set to "promiscuous-bridge"
+```
+
+* If the effective hairpin mode is `hairpin-veth`, ensure `Kubelet` has
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: the Kubelet

File: docs/user-guide/debugging-services.md [hunk 8]
```diff
@@ -555,6 +555,63 @@ There are three things to check:
   Engine doesn't do port remapping, so if your application serves on 8080,
   the `containerPort` field needs to be 8080.
 
+### A Pod cannot reach itself via Service IP
+
+This mostly happens when `kube-proxy` is running in `iptables` mode and `Pod`s
+are connected with bridge network. There is `--hairpin-mode` flag in `Kubelet`.
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: please reword: the kubelet exposes a `hairpin-mode` [flag] link to --hairpin-mode? that allows endpoints of a Service to loadbalance back to themselves if they should try to access their own Service VIP. The flag must either be set to `hairpin-veth`, or `promiscuous-bridge`.

File: docs/user-guide/debugging-services.md [hunk 9]
```diff
@@ -555,6 +555,63 @@ There are three things to check:
   Engine doesn't do port remapping, so if your application serves on 8080,
   the `containerPort` field needs to be 8080.
 
+### A Pod cannot reach itself via Service IP
+
+This mostly happens when `kube-proxy` is running in `iptables` mode and `Pod`s
```
Comment (bprashanth) [inline docs/user-guide/debugging-services.md]: s/`Pods`/Pods

**Comment by bprashanth** [comment]
LGTM just nits

**Comment by freehan** [comment]
Fixed. Merging.

**Pull request merged**